subprojects/language-java/src/main/java/org/gradle/api/tasks/JavaExec.java

```diff
@@ -743,11 +731,31 @@ private String getEffectiveExecutable() {
         if (javaLauncher.isPresent()) {
             return javaLauncher.get().getExecutablePath().toString();
         }
-        final String executable = getExecutable();
-        if (executable != null) {
-            return executable;
+
+        JavaToolchainSpec toolchain = JavaExecExecutableUtils.getExecutableOverrideToolchainSpec(this, getObjectFactory());
```

Comment:
Before this change, was `JavaExec` able to run with an executable which doesn't come from a Java installation?

Reply:
It seems it was not possible, but for an unrelated reason.

On the current master, if we specify an executable path that does not lead to a Java installation `java` binary, the build fails. This happens during snapshotting of the inputs for the `JavaExec` task. In the `JavaExec.getJavaVersion` a `JvmVersionDetector` is used. It fails if supplied with a path leading to a binary outside of a Java installation. 

Here is the usage in `JavaExec`:
https://github.com/gradle/gradle/blob/2b8957f5157135a335bbe4db9c5c54e1e89b0025/subprojects/language-java/src/main/java/org/gradle/api/tasks/JavaExec.java#L515

Here is the validation of the Java installation by the detector:
https://github.com/gradle/gradle/blob/d273aa1b79ca6600623f4eaee09bc743ad784792/subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/DefaultJvmVersionDetector.java

It seems like `DefaultJvmVersionDetector` is another candidate for consolidation of logic of Java installation verification where we do the `.getParentFile().getParentFile()` thing.